### ISSUE TITLE
Add `allowEscape` or another event to handle escape without deactivation

### ISSUE THREAD

**Issue by bpl** [focus-trap/focus-trap#126]
How do you feel about adding `allowEscape` to complement `escapeDeactivates`, similar to the `clickOutsideDeactivates`/`allowClickOutside` pair?

I have a use case where I want to navigate to a different page when the user clicks outside or hits escape inside a modal. If the user cancels navigation, the modal should remain open and the focus trap should remain in effect. If `allowEscape` existed, I could achieve the behavior with the following options:

```javascript
{
    clickOutsideDeactivates: true,
    escapeDeactivates: true,
    allowClickOutside: () => {
        window.location.href = 'https://www.example.com/';
        return true;
    },
    allowEscape: () => {
        window.location.href = 'https://www.example.com/';
        return true;
    }
}
```

Another option would be to add `onClickOutside` and `onEscape` events that are invoked regardless of `clickOutsideDeactivates` and `escapeDeactivates` properties. The configuration would be:

```javascript
{
    clickOutsideDeactivates: false,
    escapeDeactivates: false,
    onClickOutside: () => window.location.href = 'https://www.example.com/',
    onEscape: () => window.location.href = 'https://www.example.com/'
}
```

In any case, I'm happy to send a pull request with the implementation if I can get an agreement on what the API should look like.

**Edit:** Added another example.

**Comment by stefcameron** [comment]
@bpl I'm a new maintainer. I had other things to do to the repo and just catching up now. 😄  I could see adding an `allowEscape` to compliment `allowClickOutside`, however, your example of how you would support your use case wouldn't work, based on what I see in the code WRT to how `allowClickOutside` works when `clickOutsideDeactivates` is `true`, that being that `allowClickOutside` (when a function) is **not** called: https://github.com/focus-trap/focus-trap/blob/master/index.js#L228-L232

What you would need to do instead is this:
```
let focusTrap;

const clickEscHandler = (event) => {
    window.location.href = 'https://www.example.com/';
    // determine if user does NOT cancel nav somehow (that's your use
    //  case for terminating the trap, and maybe this is async, I don't know
    //  how you intend to do it)
    if (navNotCanceled) {
        focusTrap.deactivate();
    }
    return true;
};

focusTrap = createFocusTrap('#trap', {
    clickOutsideDeactivates: false,           // <- false, not true
    escapeDeactivates: true,                     // <- false, not true
    allowClickOutside: clickEscHandler,
    allowEscape: clickEscHandler,
};
```

In other words, if we add `allowEscape`, I think it should work in the same way as `allowClickOutside` and `clickOutsideDeactivates` behave as a pair.

Thoughts?

**Comment by bpl** [comment]
@stefcameron Thanks for your reply and pointing out that I wasn't reading the documentation correctly on the interaction of `clickOutsideDeactivates` and `allowClickOutside`.

Given the actual behavior, I'm wondering if pairing `allowClickOutside` with `onEscape` would be the best option? The pair of events wouldn't be nicely symmetric but the naming might better reflect the behavior since I'm not sure if there's anything to allow or disallow on escape given the actual behavior of `escapeDeactivates`.

**Comment by stefcameron** [comment]
> @stefcameron Thanks for your reply and pointing out that I wasn't reading the documentation correctly on the interaction of `clickOutsideDeactivates` and `allowClickOutside`.

My pleasure!

> Given the actual behavior, I'm wondering if pairing `allowClickOutside` with `onEscape` would be the best option? The pair of events wouldn't be nicely symmetric but the naming might better reflect the behavior since I'm not sure if there's anything to allow or disallow on escape given the actual behavior of `escapeDeactivates`.

Looking further at the [ESC handler](https://github.com/focus-trap/focus-trap/blob/master/index.js#L258-L268) vs the [click](https://github.com/focus-trap/focus-trap/blob/master/index.js#L224-L246) [handler](https://github.com/focus-trap/focus-trap/blob/master/index.js#L288-L301), I see a big difference: While the click handler will `preventDefault()` _and_ `stopImmediatePropagation()` on the click event (unless `clickOutsideDeactivates: true` or `allowOutsideClick` permits the click), the ESC handler does not. In contrast, the ESC handler does not `preventDefault()` and just does nothing, which _should_ mean that the keyboard event propagates normally.

That should mean that you could handle the keyboard event simply by setting `escapeDeactivates: false`. And then you can set the `window.location` and decide if you'll manually deactivate the trap (as in [my illustration](https://github.com/focus-trap/focus-trap/issues/126#issuecomment-687668328)).

For the click, we already have `clickOutsideDeactivates: false, allowOutsideClick: (event) => boolean` that should work in a similar way.

👉  I'm not saying there isn't room to improve the experience/elegance of the API here, but I'm wondering if, as it stands today, you could actually achieve your use case without any changes.

Now, I'm not sure what to make of `allowOutsideClick` actually being called __twice__ on every click: First on `mousedown/touchstart`, and then another time (in the same "click sequence") when the actual `click` is registered. That doesn't seem right to me. I would expect that `allowOutsideClick` would be called on `mousedown/touchstart` and the response remembered. When the subsequent `click` event takes place, check the registered response and either let the click through or not, only calling `allowOutsideClick` once.

I think that should be fixed.

One thing to note is that `allowOutsideClick` seems to be there to support a mobile touch use case, which isn't easily tested...

And now if we added `onOutsideClick(event: MouseEvent | TouchEvent)` and `onEscape(event: KeyboardEvent)`, I would say that `onOutsideClick` should be called just before [line 228](https://github.com/focus-trap/focus-trap/blob/master/index.js#L228), and `onEscape` just before [line 259](https://github.com/focus-trap/focus-trap/blob/master/index.js#L259) (only if it's ESC, of course; line 259 would need to be re-worked a bit).

But, what happens if `event.preventDefault()` is called in either one? Maybe we  interpret that as "don't deactivate on click/ESC event if `clickOutsideDeactivates/escapeDeactivates: true`"?

Then for `allowOutsideClick`, it still gets called and you can return true/false to decide if you want the click to go through. (And there's no equivalent for ESC because, AFAICT, the event _does_ go through regardless of `escapeDeactivates`.)

**Comment by Windvis** [comment]
Hey! I have a similar use case where these `onEscape` and `onClickOutside` events could be really handy. 

We created a custom modal component which uses focus-frap internally. That component calls a user provided function when the modal needs to be closed. This also needs to happen when the user presses the escape key, or clicks outside the modal body. Since focus-trap provides that functionality already we want to reuse that. At the moment we kind of misuse the `onDeactivate` event since that's also called when those user actions happen, but it's kind of a hack since we are not really interested in the "deactivate" part. 

The fact that we use `onDeactivate` also means we have to catch some other cases where our logic shouldn't be executed. So new events that are called only when those specific actions happen would solve our use-case in a better way.

**Comment by stefcameron** [comment]
@Windvis 

> Hey! I have a similar use case where these `onEscape` and `onClickOutside` events could be really handy.
> 
> We created a custom modal component which uses focus-frap internally. That component calls a user provided function when the modal needs to be closed. This also needs to happen when the user presses the escape key, or clicks outside the modal body. Since focus-trap provides that functionality already we want to reuse that. At the moment we kind of misuse the `onDeactivate` event since that's also called when those user actions happen, but it's kind of a hack since we are not really interested in the "deactivate" part.
> 
> The fact that we use `onDeactivate` also means we have to catch some other cases where our logic shouldn't be executed. So new events that are called only when those specific actions happen would solve our use-case in a better way.

Thanks for reaching out! So my proposal from [a while ago](https://github.com/focus-trap/focus-trap/issues/126#issuecomment-691526235) was basically this:

> And now if we added onOutsideClick(event: MouseEvent | TouchEvent) and onEscape(event: KeyboardEvent), I would say that onOutsideClick should be called just before line 228, and onEscape just before line 259 (only if it's ESC, of course; line 259 would need to be re-worked a bit).
>
> But, what happens if event.preventDefault() is called in either one? Maybe we interpret that as "don't deactivate on click/ESC event if clickOutsideDeactivates/escapeDeactivates: true"?

But now I'm thinking that `onOutsideClick(event: MouseEvent | TouchEvent) => boolean` is the same thing as the existing `allowOutsideClick` option when you specify a function for it instead of a boolean -- unless the use case is to use here (in your case, and in @bpl's case) is to prevent deactivation under certain circumstances, _even if_ `clickOutsideDeactivates` is `true`.

Incidentally, since last September, the `clickOutsideDeactivates` now accepts a function. The function gets the event object, and must return a boolean to let `focus-trap` know if that click should deactivate or not.

So I'm thinking that for clicks, your problem might be solved already.

And as for `onEscape()`, I'm looking at the existing `escapeDeactivates` option, and also thinking to add a function option to that so that if it's a function, we look at the return value boolean there too, and let you decide then and there if ESC should deactivate or not.

WDYT? Would that solve your current `onDeactivate()` hack?

**Comment by Windvis** [comment]
@stefcameron Yes `escapeDeactivates` also accepting a function would be perfect! Combining that with `clickOutsideDeactivates` would indeed cover my use-case!

**Comment by stefcameron** [comment]
I opened #442 to fix this.

**Comment by stefcameron** [comment]
This is being published in 6.6.0 right now.
